### ISSUE TITLE
Speed adjustment not working

### ISSUE THREAD

**Issue by darkamgine** [webosbrew/youtube-webos#6]
When using this package, changing the speed modifer doesn't seem to take effect. The video still plays at 1.0x speed no matter if 0.25x or 2.0x is selected. This happens with the friedchickenbutt version as well, but works as expected with the original unmodded version. 

Tv: lg oled b9
Webos: 05.00.15

**Comment by Informatic** [comment]
If I remember correctly this is a limitation of webOS WAM browser on webOS 3.x and 4.x. Not sure about webOS 5.x (above version is firmware version, not webOS TV version, I suspect B9 is around webOS 5.x)

**Comment by darkamgine** [comment]
Ah the b9 is tv version 4.9.0 goldilocks gyeryongsan so that makes sense. Does this mean there's no way around this issue on this model?

**Comment by Crazor** [comment]
Well, on my B9SLA, the official YouTube app does support changing the speed, albeit with audio artifacts that I don't get on my computers or my phone.
This fork does have the speed controls, but they do not do anything.

**Comment by aabytt** [comment]
at the same time playback speed works fine when running the package in lg webos emulator

**Comment by Informatic** [comment]
Just tested on webOS 3.8 with `default` `trustLevel` application (youtube is running on `netcast` due to need of user-agent spoofing) and video `playbackRate` adjustment causes audio to go mute.

The fact it works in an emulator is likely caused by discrepancies between video flow - emulator definitely uses software decoding, likely one already built into Chrome.

**Comment by Manu99it** [comment]
Just for confirming, even for me doesn't work on webOS 6.2.0-32 latest build for c1 (03.21.05), official app instead yes. Is this fixable? I use very often this feature😬

**Comment by Informatic** [comment]
Does speed adjustment in official app have working audio?

In general - this is going to be fairly hard - we'll probably need to release multiple app versions... But we'll see.

**Comment by Manu99it** [comment]
> Does speed adjustment in official app have working audio?
> 
> In general - this is going to be fairly hard - we'll probably need to release multiple app versions... But we'll see.

Yep, in official app audio works perfectly from 0.25 to 2x speed

**Comment by KTsarlz** [comment]
Wondering if anyone was able to finx a fix? On 0.3.1 the playback speed is not there on my C1, but on the official youtube app it's there

**Comment by DeMoN3rock** [comment]
Hi all,
ok, it's not possible to adjust speed, but...is it possibile to install BOTH official and non-official youtube app on the same tv?
In this way I can use the unofficial when I want to skip ads, but the official when I need to view something in 2x

PS: reVanced and similar youtube app let 2x playback and ads skipping at the same time; I dunno if the code base is similar, could be usefull to see at them for implementing this funcionality?

**Comment by throwaway96** [comment]
As far as I know, you can have this installed (and functional) at the same time as the official app if you change the app ID (`id` in `assets/appinfo.json`).

**Comment by DrA1ex** [comment]
Hi everyone!

I’ve discovered a method to adjust playback speed in WebOS, and I’ve confirmed that it works. Here’s a description of the method: [link](https://gist.github.com/aabytt/bddbb1bcf031a050d89a89aeee3a6737?permalink_comment_id=4095567#playback-speed).

### Example code:
```javascript
webOS.service.request("luna://com.webos.media", {
    method: "setPlayRate",
    parameters: { 
        'mediaId': "_tul03L7E2FoL5d",
        'playRate': 1.5,
        'audioOutput': true
    }
});
```

**Comment by aabytt** [comment]
@DrA1ex 
`setPlayRate` and other `com.webos.media` methods not available in YTAF because it uses `netcast` `trustlevel` to spoof user-agent

**Comment by DrA1ex** [comment]
> `setPlayRate` and other `com.webos.media` methods not available in YTAF because it uses `netcast` `trustlevel` to spoof user-agent

Just tried to build package with default `trustLevel` and without custom `user-agent`, but seems like it still doesn't work.
I also tried changing `playbackRate` via `ares-inspect` and that doesn't work either.

It looks like there is a way to implement playback speed (since it works in the browser or in the Lampa app), but it would require the LS2 API and custom video playback logic. No easy way, unfortunately.

**Comment by ut4utc** [comment]
Playback speed adjustment unavailable: When attempting to change the video speed, the message "Speed - Not available on this device" appears. The TV is new and powerful, so this feature should be supported.

Device - 55UR91006LA
Firmware version: 13.40.97
WebOS version: 8.4.0

**Comment by throwaway96** [comment]
As far as I know, webOS 8 has the same restrictions as previous versions. It was never about the TV's hardware capabilities.

**Comment by rpcarvalheira** [comment]
Ehhh... @throwaway96 
https://github.com/UltraHDR/youtube-webos-cobalt?tab=readme-ov-file

Maybe this could help? The guy used google cobalt as a proxy to enable speeds and solve this issue somehow.

**Comment by xTombstoneActual** [comment]
Hi, thanks for the amazing work.
Just want to update that the official youtube app does support speeed change on models such as C2 without any issue.
Is there a chance that it can be implemented now since official app support. 
Thankyou for all the Hard work.

**Comment by bestskillzzz** [comment]
> Ehhh... @throwaway96 
> https://github.com/UltraHDR/youtube-webos-cobalt?tab=readme-ov-file
> 
> Maybe this could help? The guy used google cobalt as a proxy to enable speeds and solve this issue somehow.

Does it still work if I use and patch iPK using that
Thanks much

**Comment by rpcarvalheira** [comment]
To be honest, I actually have no effing clue, but would love if it did. Would you mind trying it?

**Comment by bestskillzzz** [comment]
> To be honest, I actually have no effing clue, but would love if it did. Would you mind trying it? 

The instructions are outdated, the repository they are asking me to clone doesn't exist anymore but I can try to clone his repository but the instructions are kind of outdated and confusing so I don't think it is going to work but I can try patching the iPK file and tell you how that goes, but if you could do it that could be better as I am in a family and we are all sharing 1 computer, but usually my dad uses it a lot.

**Comment by bestskillzzz** [comment]
> To be honest, I actually have no effing clue, but would love if it did. Would you mind trying it? 

Okay looks like with the help with ai and correcting all the permission issues I had it  actually did work but I used this repository https://github.com/webos-uxomis/youtube-webos-cobalt-browser, as it was updated more recently from 9 months ago, I would attach iPK but it seems like GitHub doesn't allow you to attach IPK files

**Comment by rpcarvalheira** [comment]
I see. You may create a fork and put it as release and send the link here :)

**Comment by bestskillzzz** [comment]
> I see. You may create a fork and put it as release and send the link here :)

Ok here is link: https://github.com/Sahibishere/youtube-webos-cobalt-browser/releases/tag/YouTube-1.41.3-Patched

**Comment by bestskillzzz** [comment]
> I see. You may create a fork and put it as release and send the link here :)

Did it work for you, I tried using the official application, the youtube-webos application from this repository, and using the repository you linked earlier and I either run into a problem where playback speed isn't supported or it is supported but I get ads. It's probably because it's outdated

**Comment by iGaneshKumar** [comment]
> > I see. You may create a fork and put it as release and send the link here :)
> 
> Did it work for you, I tried using the official application, the youtube-webos application from this repository, and using the repository you linked earlier and I either run into a problem where playback speed isn't supported or it is supported but I get ads. It's probably because it's outdated

The instructions [here](https://github.com/webos-uxomis/youtube-webos-cobalt-browser) are quite confusing, so please clarify. Which YouTube app did you update? Was it the official version available on the LG Store, or the YouTube AdFree app? Also, did you uninstall the official YouTube app before installing the youtube-webos app?

**Comment by bestskillzzz** [comment]
> > > I see. You may create a fork and put it as release and send the link here :)
> > 
> > Did it work for you, I tried using the official application, the youtube-webos application from this repository, and using the repository you linked earlier and I either run into a problem where playback speed isn't supported or it is supported but I get ads. It's probably because it's outdated
> 
> The instructions [here](https://github.com/webos-uxomis/youtube-webos-cobalt-browser) are quite confusing, so please clarify. Which YouTube app did you update? Was it the official version available on the LG Store, or the YouTube AdFree app? Also, did you uninstall the official YouTube app before installing the youtube-webos app?

Yes it was the official one, I have a rooted TV so it was pretty easy to find the extracted YouTube folder, I then package it using ares and then patched that iPK, I am pretty sure they want the official version because they want the version that has cobalt which is only the official version, but there is no use doing it cause the patched ipks don't work, they didn't crash for me but it would still display ads.

**Comment by the4anoni** [comment]
> > I see. You may create a fork and put it as release and send the link here :)
> 
> Ok here is link: https://github.com/Sahibishere/youtube-webos-cobalt-browser/releases/tag/YouTube-1.41.3-Patched

For me it's just crashing on startup, LG C4 23.20.39 rooted.

**Comment by Knacki** [comment]
> > I see. You may create a fork and put it as release and send the link here :)
> 
> Ok here is link: https://github.com/Sahibishere/youtube-webos-cobalt-browser/releases/tag/YouTube-1.41.3-Patched

did not work. LG 65W7 rooted, its crashing on startup

**Comment by horst789github** [comment]
Same Problem, Not rooted. OLED77C38LA

**Comment by masmelsony** [comment]
Any new developments happened here?

**Comment by repparw** [comment]
Workaround that fixed this for me on rooted LG webOS using official YouTube Cobalt app:
- Root cause in my case was Cobalt binary compatibility when replacing libcobalt.so, not the speed UI itself.
- Building from older 23.lts artifacts caused unresolved symbol/runtime mismatch on TV loader; app either crashed or behaved like stock app.
- Reliable path was building from 25.lts.stable (SB API 16) and applying only minimal patches:
  1) Inject adblock JS/CSS in cobalt/dom/document.cc
  2) Allow sponsorblock.inf.re in CSP (cobalt/csp/directive_list.cc)
  3) Keep evergreen docker output path aligned for SB-versioned out dir
- Then use clean install cycle each test: close app -> uninstall -> install patched IPK -> launch (reboot if stale state suspected).

My fork with working patch set:
https://github.com/repparw/youtube-webos-cobalt-browser
Branch: feature/cobalt-fixes
Patch file: cobalt-patches/cobalt-25.lts.stable.patch

**Comment by horst789github** [comment]
@repparw how I can use your solution without rooting? Can we use your ipk with homebrew loader?

**Comment by repparw** [comment]
> @repparw how I can use your solution without rooting? Can we use your ipk with homebrew loader? 

I'm rooted, so I don't know. you could try I guess, just uninstall the official one and install the ipk

**Comment by masmelsony** [comment]
> Workaround that fixed this for me on rooted LG webOS using official YouTube Cobalt app:
> 
> * Root cause in my case was Cobalt binary compatibility when replacing libcobalt.so, not the speed UI itself.
> * Building from older 23.lts artifacts caused unresolved symbol/runtime mismatch on TV loader; app either crashed or behaved like stock app.
> * Reliable path was building from 25.lts.stable (SB API 16) and applying only minimal patches:
>   
>   1. Inject adblock JS/CSS in cobalt/dom/document.cc
>   2. Allow sponsorblock.inf.re in CSP (cobalt/csp/directive_list.cc)
>   3. Keep evergreen docker output path aligned for SB-versioned out dir
> * Then use clean install cycle each test: close app -> uninstall -> install patched IPK -> launch (reboot if stale state suspected).
> 
> My fork with working patch set: https://github.com/repparw/youtube-webos-cobalt-browser Branch: feature/cobalt-fixes Patch file: cobalt-patches/cobalt-25.lts.stable.patch

i have a webos 22 lgtv and i installed the ipk you provided but it doesn't work it crashes immediately for me i tried making my own version by patching an ipk but 🥲 that stuff isn't for me i couldn't for the life of me get an official youtube ipk couldn't find it in User/palm/application or media/developer/apps/user/palm/applications so i would appreciate it if you can tell me something i can do i really want that speed control it's the only reason i don't watch youtube on the tv

i know i didn't provide much usefull information but that's all i can do sorry and thanks for your time

**Comment by repparw** [comment]
rooted? I just threw a lot of tokens at the problem and gave the agent ssh access to the tv to check logs, etc. didn't do much myself.

re getting the official ipk: I had to install the app from the store and get the link from the logs and download it to the pc before the link expired.

if you're rooted, first step would be to check the logs and determine why it's crashing. good luck

Most of the work was done by gpt5.3 codex (used 5.3 because I didn't have much limit, using copilot student. probably 5.4 would do a better job)
